agent/src/agent/data_extractor/observium/observium.py

```diff
@@ -0,0 +1,103 @@
+import time
+import requests
+import urllib.parse
+
+from requests.auth import HTTPBasicAuth
+from agent import source
+from agent.modules import logger, field, lookup
+from agent.pipeline import Pipeline
+from agent.source import ObserviumSource
+
+logger_ = logger.get_logger(__name__)
+
+N_REQUESTS_TRIES = 3
+
+POLL_TIME_KEYS = {
+    'ports': 'poll_time',
+    'mempools': 'mempool_polled',
+    'processors': 'processor_polled',
+    'storage': 'storage_polled',
+}
+
+RESPONSE_DATA_KEYS = {
+    'ports': 'ports',
+    'mempools': 'entries',
+    'processors': 'entries',
+    'storage': 'storage',
+}
+
+
+@lookup.provide
+def extract_metrics(pipeline_: Pipeline) -> list:
+    base_url = urllib.parse.urljoin(pipeline_.source.config[source.ObserviumSource.URL], '/api/v0/')
+    endpoint = pipeline_.source.config['endpoint']
+    data = _get(
+        pipeline_.source,
+        urllib.parse.urljoin(base_url, endpoint),
+        pipeline_.config.get('request_params', {}),
+        RESPONSE_DATA_KEYS[endpoint],
+    )
+    data = _add_devices_data(data, pipeline_)
+    return _create_metrics(data, pipeline_)
+
+
+def _get(source_: ObserviumSource, url: str, params: dict, response_key: str):
+    for i in range(1, N_REQUESTS_TRIES + 1):
+        try:
+            res = requests.get(
+                url,
+                auth=HTTPBasicAuth(
+                    source_.config[source.ObserviumSource.USERNAME],
+                    source_.config[source.ObserviumSource.PASSWORD],
+                ),
+                params=params,
+                verify=source_.verify_ssl,
+                timeout=source_.query_timeout,
+            )
+            res.raise_for_status()
+            return res.json()[response_key]
+        except requests.HTTPError as e:
```

Comment:
we need to send source error metric to our monitoring in case of error response from observium